docs/design/query-caching.md

```diff
@@ -0,0 +1,224 @@
+- Agnostic towards GraphQL or REST request metadata.
+- Determinate size based on the size of the string generated.
+- Equivalent query metadata will map to the same cache value.
+  - This protects against reference type comparison where two distinct C# objects with same property values (request metadata) are treated as two separate cache entries instead of the same entry.
+
+- Rely on the underlying cache implementation to store the key/value pair in a concurrency compatible collection. The implementation defined collection will abstract away the work done to handle cache key hash collisions and appropriate bucketing.
+  - Important to note that we can't and won't use a hashcode as the verbatim string for a cache key. Microsoft Docs remarks on default implementation of GetHashCode() say not to use a hashcode as a key to retrieve an object from a keyed collection or to store the value in a database because the value is not “permanent.” https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-7.0#remarks
+  - If in the future we determine that we want to optimize to creating a consistently sized string to provide as a cache key, we can evaluate hashing the string generated. One example would be SHA256 hashing, however sha256 is a crypto operation and may result in a performance degradation that makes caching benefits neglible.
+
+TODO: Does collation affect the cache key?
+
+#### Value
+
+The value stored in the cache is the database response serialized as a JSON string.
+
+The size of the value stored is variable and dependent on the data returned by the database. 
+
+#### Size
+
+The maximum cache size is dependent on the memory (RAM) specifications of the hosting server. The cache will share memory space with the running instance of DAB and other operating system/hosting resources. 
+Since key/value pairs are not consistently sized, we can utilize string length (character count) to determine a relative size of the cache entry. 
+
+### Cache Behavior
+
+#### Eviction
+
+Cache entries are evicted when the cache gets close to our predetermined size limit, and when entries expire based on their configured ttl.
+
+#### Updates
+
+Detect if the operation is a follow up read to a mutation. We can then map that follow up read generated query to a cache key entry and attempt to invalidate the cache entry. This would not allow us to expire all variations of read queries generated for the specific entity(s) because querystrings, role based access, and column based access may be cached with different keys. This behavior backs up our guidance to limit caching to rarely changing database objects/records because cache entries may be stale and fed back to the user.
+
+## Security and privacy
+
+### Data residency
+
+__Cache data is stored in-memory__ and remains in memory until evicted due to ttl rules, manual eviction, or DAB process shutdown. Cache data is not persisted to disk.
```

Comment:
Are there security concerns with holding information from the database locally in memory? Was just thinking about the possibility of potentially sensitive information living in memory for too long, potentially creating a vulnerability.

Reply:
This is why cache config is per-entity. Entities with more sensitive fields can choose whether to opt in to using the cache. 

What vulnerabilities came to mind? If there is a compromise of the server hosting dab, then that's already a bigger problem. 

One case that is worth considering is either documenting or identifying a mitigating action to handle:
- database user represented by the connection string has a change of permissions, possibly has permissions removed, this should result in a purge of the cache which could then be rebuilt using most recent permissions. This would require event notification to dab, if one exists. 
- if database connection goes down, the cache is meant to kick in. Users who only want the db results in memory for a certain amount of time after the db goes down can configure their cache entry ttl accordingly.

Reply:
Ya, similar to the cases you describe my immediate concerns would be easy to describe scenario like some information that should not have been exposed on the database is exposed, or a policy for how the data is exposed changes, but dab has cached that information, so now the behavior between sending the query to the backend database yourself and what you see from dab is different, and in a way where dab is exposing information that the database does not want to. Maybe some way to flush the cache from the user would alleviate this?

Reply:
>scenario like some information that should not have been exposed on the database is exposed

I'm not sure i follow. Is the scenario you mean one where the connection string is granted too much access, then the cache fills with that unintentionally exposed data, and then connection string access is fixed, and now the cache has data that should not be accessible?
In that case, e
- dab should have mechanism to purge cache when connection string changes. 
- data in the cache was allowed to be accessed and read at the time of access. If the protection level changes at the database level, users will not be able to get new fresh data. and are subject to the stale data. That's where  the developer will need to carefully manage their infrastructure.
 
> a policy for how the data is exposed changes

A policy in the dab runtime config permissions? in this case, all authN and authZ occur before cache access occurs. If a user meets all permissions requirements, their GraphQL or REST request will be translated into a SQL query and subsequently have the query executed against the database. cache access should be allowed then too. 

> some way to flush the cache from the user

We shouldn't allow a request to dab to clear the cache. the protection mechanisms that would need to be built to accomplish this would be covered by the developer's ability to restart dab.

Reply:
> I'm not sure i follow. Is the scenario you mean one where the connection string is granted too much access, then the cache fills with that unintentionally exposed data, and then connection string access is fixed, and now the cache has data that should not be accessible?

I was thinking something more simple. Consider a database that has mistakenly exposed users social security numbers into a column that can be read, but should not be so. Someone uses dab to connect to and read this information. Then the database is corrected to not expose this information. But dab will still have the information in its cache until the cache is updated to reflect the new state of the database.

Granted, a bad actor could have saved that information on their own, so perhaps this isn't a scenario we should worry about too much, but wanted to make sure it is something that is considered. 

Reply:
Let's say a user is hosting DAB in their own hosted setup and they are able to take a memory dump in that server. In this case, would they be able to in some way see the contents of the cached responses (that would essentially enable them to read contents of a database that they might not necessarily have access to).

[I acknowledge I'm not too familiar with dumps and could be something that we may not have to worry about. But, just wanted to bring it up so that we've thought through this possibility]

Reply:
Your having a level of access to the host machine such that you can acquire a memory dump means that you can see all DAB metadata such as the connection string, which you could use to access the db directly instead of by looking at cache data. The host machine should be hardened such that only permitted users can access for production related maintenance/troubleshooting.